In [1]:
import pandas as pd
%matplotlib inline   
import seaborn as sns
import random as random


sns.set_theme()
sns.set_context("notebook")
%load_ext autoreload
%autoreload 2

In [2]:
dtype_dict = {
    'SE_Number': 'str',
    'FarmName_Pseudo': 'str',
    'CalvingDate': 'str',
    'CalvingSireBullID': 'str',
    'CalvingEase': 'str',
    'BreedName': 'str',
    'BirthDate': 'str',
    'Mother': 'str',
    'Father': 'str',
    'YearSeason': 'str',
    'MeanTemperature': 'float',
    'MeanRelativeHumidity': 'float',
    'MeanTHI_adj': 'float',
    'HW': 'Int64',
    'cum_HW': 'Int64',
    'MaxTemp15Threshold': 'Int64'
}

# Load the data using the dtype_dict
data = pd.read_csv('../Data/MergedData/TheCalvingData.csv', dtype=dtype_dict)

# Convert the date columns to datetime
date_columns = [
    'CalvingDate',
    'BirthDate'
]

for column in date_columns:
    data[column] = pd.to_datetime(data[column], errors='coerce')

data.head()

SE_Number FarmName_Pseudo CalvingDate CalvingSireBullID CalvingEase  \
0  SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman          11   
1  SE-30dc5787-1389        a624fb9a  2022-04-11   22-8352 Ventura          11   
2  SE-30dc5787-1389        a624fb9a  2023-05-15   22-8352 Ventura          11   
3  SE-30dc5787-1396        a624fb9a  2022-08-31     3-7725 Attack          11   
4  SE-a624fb9a-1039        a624fb9a  2022-01-07     3-7725 Attack          12   

  YearSeason BreedName  BirthDate            Mother Father  MeanTemperature  \
0     2022-2    02 SLB 2013-07-12               NaN    NaN        12.987500   
1     2022-2    01 SRB 2015-01-27               NaN    NaN         4.208333   
2     2023-2    01 SRB 2015-01-27               NaN    NaN        15.045833   
3     2022-3    02 SLB 2015-03-20               NaN    NaN        11.041667   
4     2022-1    02 SLB 2010-09-23  SE-a624fb9a-0966    NaN       -10.733333   

   MeanRelativeHumidity  MeanTHI_adj  HW  cum_HW  MaxTemp15Threshold  
0              0.755542    55.218420   0       0                   1  
1              0.485667    45.716935   0       0                   0  
2              0.524333    60.559813   0       0                   1  
3              0.794792    53.536477   0       0                   0  
4              0.880792    17.822442   0       0                   0

In [3]:
dtype_dict = {
    'Date': 'str',
    'FarmName_Pseudo': 'str',
    'SE_Number': 'str',
    'Age': 'Int64',
    'DailyYield': 'float',
    'PreviousDailyYield': 'float',
    'DailyYieldChange': 'float',
    'DaysInMilk': 'Int64',
    'YearSeason': 'str',
    'LactationNumber': 'Int64',
    'ExpectedYield': 'float',
    'NormalizedDailyYield': 'float',
    'NormalizedDailyYieldChange': 'float',
    'HeatStress': 'Int64',
    'Temp15Threshold': 'Int64',
    'HW': 'Int64',
    'cum_HW': 'Int64',
    'MeanTemperature': 'float',
    'MeanTHI_adj': 'float'
}

milk_data = pd.read_csv('../Data/MergedData/MilkApproachYieldData.csv', dtype=dtype_dict)
milk_data['Date'] = pd.to_datetime(milk_data['Date'], format='%Y-%m-%d')
milk_data.head(-5)

Date FarmName_Pseudo           SE_Number   Age   BreedName  \
0      2022-01-01        a624fb9a    SE-064c0cec-1189  3095      02 SLB   
1      2022-01-02        a624fb9a    SE-064c0cec-1189  3096      02 SLB   
2      2022-01-03        a624fb9a    SE-064c0cec-1189  3097      02 SLB   
3      2022-01-04        a624fb9a    SE-064c0cec-1189  3098      02 SLB   
4      2022-01-05        a624fb9a    SE-064c0cec-1189  3099      02 SLB   
...           ...             ...                 ...   ...         ...   
483097 2023-06-03        f454e660  SE-fcdf259d-0044-0  4150  41 Fjällko   
483098 2023-06-04        f454e660  SE-fcdf259d-0044-0  4151  41 Fjällko   
483099 2023-06-05        f454e660  SE-fcdf259d-0044-0  4152  41 Fjällko   
483100 2023-06-06        f454e660  SE-fcdf259d-0044-0  4153  41 Fjällko   
483101 2023-06-07        f454e660  SE-fcdf259d-0044-0  4154  41 Fjällko   

        LactationNumber  DaysInMilk YearSeason  DailyYield  \
0                     7         191     2022-1       30.77   
1                     7         192     2022-1       48.22   
2                     7         193     2022-1       30.53   
3                     7         194     2022-1       42.26   
4                     7         195     2022-1       38.49   
...                 ...         ...        ...         ...   
483097               10         347     2023-3       12.67   
483098               10         348     2023-3       22.31   
483099               10         349     2023-3       12.84   
483100               10         350     2023-3        9.47   
483101               10         351     2023-3        8.97   

        PreviousDailyYield  DailyYieldChange  ExpectedYield  \
0                 0.000000          0.000000      35.914865   
1                30.770000          8.725000      35.799613   
2                39.495000         -2.988333      35.684360   
3                36.506667          1.438333      35.569108   
4                37.945000          0.109000      35.453856   
...                    ...               ...            ...   
483097           14.652000         -0.622000      13.608593   
483098           14.030000          0.954000      13.516773   
483099           14.984000         -0.092000      13.424952   
483100           14.892000         -0.284000      13.333131   
483101           14.608000         -1.356000      13.241310   

        NormalizedDailyYield  NormalizedDailyYieldChange  HeatStress  \
0                   0.856748                    0.000000           0   
1                   1.103224                    0.243718           0   
2                   1.023044                   -0.083744           0   
3                   1.066796                    0.040438           0   
4                   1.073339                    0.003074           0   
...                      ...                         ...         ...   
483097              1.030966                   -0.045706           0   
483098              1.108549                    0.070579           0   
483099              1.109278                   -0.006853           0   
483100              1.095617                   -0.021300           0   
483101              1.000807                   -0.102407           0   

        Temp15Threshold  HW  cum_HW  MeanTemperature  MeanTHI_adj  
0                     0   0       0        -3.025000    28.012944  
1                     0   0       0        -0.279167    32.898193  
2                     0   0       0         2.033333    36.760487  
3                     0   0       0         0.066667    31.939524  
4                     0   0       0        -3.700000    26.498206  
...                 ...  ..     ...              ...          ...  
483097                1   0       0        12.666667    53.132530  
483098                1   0       0        13.079167    56.726870  
483099                1   0       0        14.237500    58.482418  
483100                1   0       0        15.345833    60.546358

In [4]:
# Merge the data based on Date and FarmName_Pseudo
merged_data = pd.merge(data, milk_data[['Date', 'FarmName_Pseudo', 'HeatStress']],
                       left_on=['CalvingDate', 'FarmName_Pseudo'],
                       right_on=['Date', 'FarmName_Pseudo'],
                       how='left')

# Drop the extra 'Date' column from the merge
merged_data = merged_data.drop(columns=['Date'])

merged_data.head(-5)

SE_Number FarmName_Pseudo CalvingDate CalvingSireBullID  \
0       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
1       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
2       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
3       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
4       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
...                  ...             ...         ...               ...   
336846  SE-afdd9a78-1209        afdd9a78  2022-09-03           09-6247   
336847  SE-afdd9a78-1210        afdd9a78  2022-09-30           03-9916   
336848  SE-afdd9a78-1211        afdd9a78  2022-08-22           09-8824   
336849  SE-afdd9a78-1215        afdd9a78  2022-10-24           09-8817   
336850  SE-afdd9a78-1217        afdd9a78  2022-12-13           09-8824   

       CalvingEase YearSeason BreedName  BirthDate            Mother   Father  \
0               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
1               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
2               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
3               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
4               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
...            ...        ...       ...        ...               ...      ...   
336846          11     2022-4    01 SRB 2020-03-18  SE-afdd9a78-1138  09-6225   
336847          11     2022-4    01 SRB 2020-03-21  SE-afdd9a78-1139  09-6225   
336848          11     2022-3    01 SRB 2020-03-22  SE-afdd9a78-1106  09-6223   
336849          11     2022-4    01 SRB 2020-06-15  SE-afdd9a78-1099  03-9862   
336850          11     2022-1    01 SRB 2020-06-15  SE-afdd9a78-1087  09-6225   

        MeanTemperature  MeanRelativeHumidity  MeanTHI_adj  HW  cum_HW  \
0             12.987500              0.755542    55.218420   0       0   
1             12.987500              0.755542    55.218420   0       0   
2             12.987500              0.755542    55.218420   0       0   
3             12.987500              0.755542    55.218420   0       0   
4             12.987500              0.755542    55.218420   0       0   
...                 ...                   ...          ...  ..     ...   
336846         9.708333              0.975125    51.746774   0       0   
336847         7.700000              0.843458    48.796863   0       0   
336848        14.570833              0.769917    60.174253   1       0   
336849         6.975000              0.961333    42.482672   0       0   
336850       -10.441667              0.816000    17.458285   0       0   

        MaxTemp15Threshold  HeatStress  
0                        1           0  
1                        1           0  
2                        1           0  
3                        1           0  
4                        1           0  
...                    ...         ...  
336846                   0        <NA>  
336847                   0        <NA>  
336848                   1        <NA>  
336849                   0        <NA>  
336850                   0        <NA>  

[336851 rows x 17 columns]

In [5]:
# Define the function to identify heat stress based on weather conditions
def identify_weather_based_heat_stress(row):
    if (row['HW'] == 1 or 
        row['cum_HW'] > 0 or 
        row['MeanTHI_adj'] > 61 or
        row['MeanTemperature'] > 15):
        return 1
    return 0

# Apply the function to fill NaN values in the HeatStress column
merged_data['HeatStress'] = merged_data.apply(
    lambda row: identify_weather_based_heat_stress(row) if pd.isna(row['HeatStress']) else row['HeatStress'], axis=1)

merged_data.head(-5)

SE_Number FarmName_Pseudo CalvingDate CalvingSireBullID  \
0       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
1       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
2       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
3       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
4       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
...                  ...             ...         ...               ...   
336846  SE-afdd9a78-1209        afdd9a78  2022-09-03           09-6247   
336847  SE-afdd9a78-1210        afdd9a78  2022-09-30           03-9916   
336848  SE-afdd9a78-1211        afdd9a78  2022-08-22           09-8824   
336849  SE-afdd9a78-1215        afdd9a78  2022-10-24           09-8817   
336850  SE-afdd9a78-1217        afdd9a78  2022-12-13           09-8824   

       CalvingEase YearSeason BreedName  BirthDate            Mother   Father  \
0               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
1               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
2               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
3               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
4               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
...            ...        ...       ...        ...               ...      ...   
336846          11     2022-4    01 SRB 2020-03-18  SE-afdd9a78-1138  09-6225   
336847          11     2022-4    01 SRB 2020-03-21  SE-afdd9a78-1139  09-6225   
336848          11     2022-3    01 SRB 2020-03-22  SE-afdd9a78-1106  09-6223   
336849          11     2022-4    01 SRB 2020-06-15  SE-afdd9a78-1099  03-9862   
336850          11     2022-1    01 SRB 2020-06-15  SE-afdd9a78-1087  09-6225   

        MeanTemperature  MeanRelativeHumidity  MeanTHI_adj  HW  cum_HW  \
0             12.987500              0.755542    55.218420   0       0   
1             12.987500              0.755542    55.218420   0       0   
2             12.987500              0.755542    55.218420   0       0   
3             12.987500              0.755542    55.218420   0       0   
4             12.987500              0.755542    55.218420   0       0   
...                 ...                   ...          ...  ..     ...   
336846         9.708333              0.975125    51.746774   0       0   
336847         7.700000              0.843458    48.796863   0       0   
336848        14.570833              0.769917    60.174253   1       0   
336849         6.975000              0.961333    42.482672   0       0   
336850       -10.441667              0.816000    17.458285   0       0   

        MaxTemp15Threshold  HeatStress  
0                        1           0  
1                        1           0  
2                        1           0  
3                        1           0  
4                        1           0  
...                    ...         ...  
336846                   0           0  
336847                   0           0  
336848                   1           1  
336849                   0           0  
336850                   0           0  

[336851 rows x 17 columns]

In [6]:
# Save the reordered DataFrame to a CSV file
merged_data.to_csv('../Data/MergedData/HeatStressCalvingData.csv', index=False)

### Variables Explanation for `HeatStressCalvingData.csv`

1. **SE_Number**:
   - Description: A unique identifier for each cow.
   - Datatype: `str`
   - Example: `SE-064c0cec-1189`

2. **FarmName_Pseudo**:
   - Description: A pseudonym for the farm where the data was collected.
   - Datatype: `str`
   - Example: `a624fb9a`

3. **CalvingDate**:
   - Description: The date when the cow calved.
   - Datatype: `str` (should be converted to `datetime` for operations)
   - Format: `YYYY-MM-DD`
   - Example: `2022-05-25`

4. **CalvingSireBullID**:
   - Description: The identifier for the bull used during the calving.
   - Datatype: `str`
   - Example: `9-4733 Fillman`

5. **CalvingEase**:
   - Description: A categorical variable indicating the ease of calving.
   - Datatype: `str`
   - Example: `11` (Easy, without assistance)

6. **BreedName**:
   - Description: The breed of the cow.
   - Datatype: `str`
   - Example: `02 SLB`

7. **BirthDate**:
   - Description: The birth date of the cow.
   - Datatype: `str` (should be converted to `datetime` for operations)
   - Format: `YYYY-MM-DD`
   - Example: `2013-07-12`

8. **Mother**:
   - Description: The identifier for the cow's mother.
   - Datatype: `str`
   - Example: `SE-a624fb9a-0966`

9. **Father**:
   - Description: The identifier for the cow's father.
   - Datatype: `str`
   - Example: `3-9071 Apollo`

10. **YearSeason**:
    - Description: The seasonal period based on the year and the month range.
    - Datatype: `str`
    - Example: `2022-2`
    - YearSeason parameters:
      - 1: Dec-Feb
      - 2: Mar-May
      - 3: Jun-Aug
      - 4: Sep-Nov

11. **MeanTemperature**:
    - Description: The mean temperature recorded on the calving day.
    - Datatype: `float`
    - Example: `12.987499999999999`

12. **MeanRelativeHumidity**:
    - Description: The mean relative humidity recorded on the calving day.
    - Datatype: `float`
    - Example: `0.7555416666666667`

13. **MeanTHI_adj**:
    - Description: The mean adjusted Temperature-Humidity Index for the calving day.
    - Datatype: `float`
    - Example: `55.21842`

14. **HW**:
    - Description: A binary variable indicating the presence of a heatwave on the calving day (0: no heatwave, 1: heatwave).
    - Datatype: `Int64`
    - Example: `0`

15. **cum_HW**:
    - Description: Cumulative number of heatwave days up to the calving day.
    - Datatype: `Int64`
    - Example: `0`

16. **MaxTemp15Threshold**:
    - Description: A binary variable indicating if the temperature exceeded 15°C on the calving day (0: no, 1: yes).
    - Datatype: `Int64`
    - Example: `1`

17. **HeatStress**:
    - Description: A binary variable indicating the presence of heat stress on the cow on the calving day, based on specific weather conditions.
    - Datatype: `Int64`
    - Example: `0`